### **1. Import Image Dataset**
from file `ISIC_2017_Data_GroundTruth_Classification.csv`, get the value of `melanoma` corresponding for each `image_id` inside directory `PROJECT_Data` only

#### **1.1 Read the data**

In [1]:
import pandas as pd

df = pd.read_csv('ISIC_2017_Data_GroundTruth_Classification.csv')
df.head()

image_id  melanoma  seborrheic_keratosis
0  ISIC_0000000       0.0                   0.0
1  ISIC_0000001       0.0                   0.0
2  ISIC_0000002       1.0                   0.0
3  ISIC_0000003       0.0                   0.0
4  ISIC_0000004       1.0                   0.0

In [2]:
import os

dataset_path = 'Z:/per-coolyeah-an/pengolahan-citra-dan-visi-komputer/Image_Processing/uas-pcvk/PROJECT_Data' # enih pathnya diganti

ids = sorted(list(set([filename[:12] for filename in os.listdir(dataset_path)]))) # ngok
len(ids)

200

#### **1.2 Data augmentation**
- rotate 15 deg
- rotate 345 deg

In [10]:
import cv2

def augment_images(dataset_path, image_id):
    image = cv2.imread(f'{dataset_path}/{image_id}.jpg')
    seg   = cv2.imread(f'{dataset_path}/{image_id}_segmentation.png')
    output_path = f'{dataset_path}/../PROJECT_Data_Augmented/{image_id}'

    def rotate_image(image, angle):
        (h, w) = image.shape[:2]
        center = (w // 2, h // 2)
        matrix = cv2.getRotationMatrix2D(center, angle, 1.0)
        return cv2.warpAffine(image, matrix, (w, h))

    rotated_image = rotate_image(image, 15)
    rotated_seg   = rotate_image(seg, 15)
    cv2.imwrite(f'{output_path}_r15.jpg', rotated_image)
    cv2.imwrite(f'{output_path}_r15_segmentation.png', rotated_seg)
    
    rotated_image = rotate_image(image, 345)
    rotated_seg   = rotate_image(seg, 345)
    cv2.imwrite(f'{output_path}_r345.jpg', rotated_image)
    cv2.imwrite(f'{output_path}_r345_segmentation.png', rotated_seg)

In [11]:
for image_id in ids:
    augment_images(dataset_path, image_id)

#### **1.3 Extract Features and Save as CSV**
calculate value of:
1. symmetry
2. reciprocal circularity (1/circularity)
3. std deviation of RGB value
   
for each `image_id` inside `PROJECT_Data`, then save as `dataset.csv` 

In [5]:
def get_features(dataset_path, image_id):
    import numpy as np
    
    segmented = cv2.cvtColor(cv2.imread(f'{dataset_path}/{image_id}_segmentation.png'), cv2.COLOR_BGR2GRAY)
    image = cv2.imread(f'{dataset_path}/{image_id}.jpg')

    contours, _ = cv2.findContours(segmented, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
    contour = max(contours, key=cv2.contourArea)

    def calc_symmetry(segmented_image):
        M = cv2.moments(segmented_image)
        cX = int(M["m10"] / M["m00"])
        cY = int(M["m01"] / M["m00"])
        
        top_left = segmented_image[:cY, :cX]
        top_right = segmented_image[:cY, cX:]
        bottom_left = segmented_image[cY:, :cX]
        bottom_right = segmented_image[cY:, cX:]
        
        symmetry_score = (cv2.matchShapes(top_left, bottom_right, 1, 0.0) + 
                        cv2.matchShapes(top_right, bottom_left, 1, 0.0)) / 2
        return { 'symmetry': symmetry_score }
    
    def calc_reciprocal_circularity(contour):
        perimeter = cv2.arcLength(contour, True)
        area = cv2.contourArea(contour)
        return { '1/circularity': perimeter**2 / (4 * np.pi * area) }

    def calc_std_color(image, segmented_image):
        masked_image = cv2.bitwise_and(image, image, mask=segmented_image)
        color_stddev = np.std(masked_image[segmented_image > 0], axis=0)
        return { 'std_color': np.mean(color_stddev) }

    features = {}

    # features.update(calc_hsv(image, segmented))
    features.update(calc_symmetry(segmented))
    features.update(calc_std_color(image, segmented))
    features.update(calc_reciprocal_circularity(contour))

    return features

In [6]:
features = {
    'image_id': [],
    'melanoma': []
}

for id in ids:
    img_features = get_features(dataset_path, id)
    img_features['image_id'] = id
    img_features['melanoma'] = int(df[df['image_id'] == id].iloc[0]['melanoma'])
    for feature_name in img_features:
        if feature_name not in features:
            features[feature_name] = []
        features[feature_name].append(img_features[feature_name])

fd = pd.DataFrame(features)

In [7]:
fd.to_csv('features.csv')
fd.head()

image_id  melanoma  symmetry  std_color  1/circularity
0  ISIC_0000000         0  0.008072  39.369340       1.463696
1  ISIC_0000001         0  0.001034  25.947679       1.716384
2  ISIC_0000002         1  0.003091  27.861042       1.973328
3  ISIC_0000003         0  0.003943  33.612228       1.465988
4  ISIC_0000004         1  0.002960  34.173329       1.202164

### **2. Train AI Classifier**
using `dataset.csv`, build an AI Classifier to predict `melanoma` (either 1 or 0) using value of 
1. asymmetry
2. border_irregularity
3. color_irregularity
4. differential_structure

#### **2.1 Split and augment the dataset**
split into test and train dataset, then augment the train data

In [8]:
from sklearn.model_selection import train_test_split
import pandas as pd

df = pd.read_csv('features.csv')
y = df['melanoma']

df_train, df_test, y_train, y_test = train_test_split(df, y, test_size=0.2, random_state=42)

X_test  = df_test[['symmetry','std_color','1/circularity']]
X_train = df_train[['symmetry','std_color','1/circularity']]

In [12]:
aug_path  = f'{dataset_path}/../PROJECT_Data_Augmented'
aug_X_train = []
aug_y_train = []

for index, row in df_train.iterrows():
    for aug in ['r15', 'r345']:
        aug_features = get_features(aug_path, f'{row['image_id']}_{aug}')

        aug_X_train.append(aug_features)
        aug_y_train.append(row['melanoma'])

In [13]:
temp_X = pd.DataFrame.from_records(aug_X_train)
temp_y = pd.Series(aug_y_train)

X_train = pd.concat([X_train, temp_X], ignore_index=True)
y_train = pd.concat([y_train, temp_y], ignore_index=True)

In [18]:
assert len(X_train) == len(y_train)
print("augmented train data length: ", len(X_train))

augmented train data length:  480


#### **2.1 Machine Learning**
ML Model for classification used in this repository: 
- Logistic Regression
- Support Vector Machine (SVM) Classifier
- K-Nearest Neighbor (KNN) Classifier

In [21]:
# NOTE: mending di run di google.colab, upload 'dataset.csv' yg udah diekstrak

import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, accuracy_score

lr_model = LogisticRegression()
lr_model.fit(X_train, y_train)

y_pred = lr_model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')

print(classification_report(y_test, y_pred))

Accuracy: 0.70
              precision    recall  f1-score   support

           0       0.74      0.67      0.70        21
           1       0.67      0.74      0.70        19

    accuracy                           0.70        40
   macro avg       0.70      0.70      0.70        40
weighted avg       0.70      0.70      0.70        40



In [15]:
from sklearn.svm import SVC

svm_model = SVC(kernel='linear', C=1.0)

svm_model.fit(X_train, y_train)

y_pred = svm_model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')
print(classification_report(y_test, y_pred))

Accuracy: 0.65
              precision    recall  f1-score   support

           0       0.67      0.67      0.67        21
           1       0.63      0.63      0.63        19

    accuracy                           0.65        40
   macro avg       0.65      0.65      0.65        40
weighted avg       0.65      0.65      0.65        40



In [16]:
from sklearn.neighbors import KNeighborsClassifier

knn_model = KNeighborsClassifier(n_neighbors=20)
knn_model.fit(X_train, y_train)

y_pred = knn_model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')
print(classification_report(y_test, y_pred))

Accuracy: 0.68
              precision    recall  f1-score   support

           0       0.70      0.67      0.68        21
           1       0.65      0.68      0.67        19

    accuracy                           0.68        40
   macro avg       0.68      0.68      0.67        40
weighted avg       0.68      0.68      0.68        40



#### **3.2 Neural Network (Multi Layer Percepteron)**
a simple multi-layer percepteron (MLP):
1. 3 dense layer, all using ReLU as activation function
    - 16 percepteron
    - 8 percepteron
    - 8 percepteron
2. a layer with 1 percepteron for classification, using sigmoid to squash the probability as activation function

In [19]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import classification_report
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.losses import BinaryCrossentropy


scaler = StandardScaler()
X_train_ft = scaler.fit_transform(X_train)
X_test_ft = scaler.transform(X_test)

model = Sequential()
model.add(Dense(16, input_dim=X_train_ft.shape[1], activation='relu'))
model.add(Dense(8, activation='relu'))
model.add(Dense(8, activation='relu'))
model.add(Dense(1, activation='sigmoid'))

model.compile(optimizer=Adam(learning_rate=0.001), 
              loss=BinaryCrossentropy(), 
              metrics=['accuracy'])

model.fit(X_train_ft, y_train, epochs=50, batch_size=10, validation_split=0.2)

loss, accuracy = model.evaluate(X_test_ft, y_test)
print(f"Test Accuracy: {accuracy}")

y_pred_prob = model.predict(X_test_ft)
y_pred = (y_pred_prob > 0.5).astype(int) # probabilitas -> klasifikasi

print(classification_report(y_test, y_pred))

c:\Python312\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/50
39/39 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.5377 - loss: 0.7063 - val_accuracy: 0.5417 - val_loss: 0.6808
Epoch 2/50
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.5901 - loss: 0.6889 - val_accuracy: 0.6250 - val_loss: 0.6624
Epoch 3/50
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.6167 - loss: 0.6593 - val_accuracy: 0.6250 - val_loss: 0.6515
Epoch 4/50
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.6622 - loss: 0.6315 - val_accuracy: 0.6458 - val_loss: 0.6425
Epoch 5/50
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.6913 - loss: 0.6233 - val_accuracy: 0.6562 - val_loss: 0.6326
Epoch 6/50
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.6707 - loss: 0.6120 - val_accuracy: 0.6771 - val_loss: 0.6227
Epoch 7/50
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.6980 - loss: 0.5959 - val_accuracy: 0.6875 - val_loss: 0.6134
Epoch 8/50
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.7139 - loss: 0.5853 - val_accuracy: 0.6562 - val_loss:

### **3. Summary**

#### **3.1 Model Comparison**
Comparison of all models shown in table below:

| Model              | Precision | Accuracy | Recall  | F1 Score |
|--------------------|-----------|----------|---------|----------|
| Logistic Regression| 0.70      | 0.70     | 0.70    | 0.70     |
| SVM                | 0.65      | 0.65     | 0.65    | 0.65     |
| KNN                | 0.68      | 0.68     | 0.68    | 0.67     |
| MLP                | 0.70      | 0.68     | 0.68    | 0.67     |

#### **3.2 Feature Extractions**
the feature needed to be extracted from image is
1. `symmetry`
    > Calculated by splitting the segmented image into four with the center is centroid of the lesions. then match the crossing region (top-left to down-right; top-right to down-left) with `cv2.matchShapes()`.
2. `std_color`
    > Done by calculating all standard deviation for every color channel (RGB) in the image then averaging the values. The image must be segmented and masked by the segmentation.
3. `1/circularity`
    > Done by calculating circularity (4pi*area)/perimeter^2 but the numerator and denominator is flipped

#### **3.3 Model Training**
The training of the model is done by splitting the dataset for train and testing with ratio of 0.2 (160 for training and 40 for testing). The 160 data for training then augmented by rotating it 15 degree and 345 degree, thus tripling the data into 480. All the features (`symmetry`, `std_color`, and `1/circularity`) in the train and test is calculated. Then the model is trained using the train data (including the augmented data).

#### **3.4 Model Pipeline**
the process to "predict" the skin lesion image with the model
1. Image acquisition
2. Skin lesions segmentation
3. Calculate value of `symmetry`, `std_color`, and `1/circularity` using the original and segmented image as a mask
4. Treat the values as feature vectors then predict with the model



In [23]:
import joblib

classifier_model = {
    'LogisticRegression': lr_model,
    'SVM': svm_model,
    'KNN': knn_model,
    'MLP': model
}

for model_name in classifier_model:
    joblib.dump(classifier_model[model_name], f'models/{model_name}.joblib')

### **4. Appendix**

repository of the project: https://github.com/Adhisetama/uas-pcvk